sycl/plugins/Intel_level0/pi_level0.hpp

```diff
@@ -0,0 +1,345 @@
+        ZeDeviceComputeProperties{} {
+    // NOTE: one must additionally call initialize() to complete
+    // PI device creation.
+  }
+
+  // Initialize the entire PI device.
+  pi_result initialize();
+
+  // L0 device handle.
+  ze_device_handle_t ZeDevice;
+
+  // PI platform to which this device belongs.
+  pi_platform Platform;
+
+  // Immediate L0 command list for this device, to be used for initializations.
+  // To be created as:
+  // - Immediate command list: So any command appended to it is immediately
+  //   offloaded to the device.
+  // - Synchronous: So implicit synchronization is made inside the level-zero
+  //   driver.
+  ze_command_list_handle_t ZeCommandListInit;
+
+  // Indicates if this is a root-device or a sub-device.
+  // Technically this information can be queried from a device handle, but it
+  // seems better to just keep it here.
+  //
+  bool IsSubDevice;
+
+  // L0 doesn't do the reference counting, so we have to do.
+  // Must be atomic to prevent data race when incrementing/decrementing.
+  std::atomic<pi_uint32> RefCount;
+
+  // Create a new command list for executing on this device.
+  // It's caller's responsibility to remember and destroy the created
+  // command list when no longer needed.
+  //
+  pi_result createCommandList(ze_command_list_handle_t *ze_command_list);
+
+  // Cache of the immutable device properties.
+  ze_device_properties_t ZeDeviceProperties;
+  ze_device_compute_properties_t ZeDeviceComputeProperties;
+};
+
+struct _pi_context {
+  _pi_context(pi_device Device)
+      : Device{Device}, RefCount{1}, ZeEventPool{nullptr},
+        NumEventsAvailableInEventPool{}, NumEventsLiveInEventPool{} {}
+
+  // L0 does not have notion of contexts.
+  // Keep the device here (must be exactly one) to return it when PI context
+  // is queried for devices.
+  //
+  pi_device Device;
+
+  // L0 doesn't do the reference counting, so we have to do.
+  // Must be atomic to prevent data race when incrementing/decrementing.
+  std::atomic<pi_uint32> RefCount;
+
+  // Get index of the free slot in the available pool. If there is no avialble
+  // pool then create new one.
+  ze_result_t getFreeSlotInExistingOrNewPool(ze_event_pool_handle_t &,
+                                             size_t &);
+
+  // If event is destroyed then decrement number of events living in the pool
+  // and destroy the pool if there are no alive events.
+  ze_result_t decrementAliveEventsInPool(ze_event_pool_handle_t pool);
+
+private:
+  // Following member variables are used to manage assignment of events
+  // to event pools.
+  // TODO: These variables may be moved to pi_device and pi_platform
+  // if appropriate
+  // Event pool to which events are being added to
+  ze_event_pool_handle_t ZeEventPool;
+  // This map will be used to determine if a pool is full or not
+  // by storing number of empty slots available in the pool
+  std::map<ze_event_pool_handle_t, pi_uint32> NumEventsAvailableInEventPool;
```

Comment:
If ordering of map entries is insignificant (which seems to be the case), unordered_map should be be more efficient

Reply:
Thx, fixed.